text/0000-oracles.md

```diff
@@ -0,0 +1,127 @@
+
+Scenario 2 (publisher-consumer):
+
+1. Off-chain Oracle worker receive messages about hurricane from sensors
+2. Worker calls `ExecuteTrigger` instruction on-chain to start benefits transferring
+
+Scenario 3 (request-response):
+
+1. End-user calls trigger T1 using existing `ExecuteTrigger` instruction
+2. Trigger T1 calls `EmitOracleRequest` instruction, providing payload with data like oracle id, request, user data, asset id to write response into and etc. That's all what this trigger does.
+3. Off-chain oracle worker receives the `OracleRequest` event, retrieves user data, makes a request to a University "X" server, and writes the response to an asset specified in original request.
+4. Trigger T2 being triggered on that asset modification. It reads the response from the asset, identifies if end-user is actually a student of University "X" and then grants a permission if so.
+
+
+
+So technically Iroha 2 already supports first two scenarios. This document aims to find a solution for the third scenario.
+
+### Implementing timeouts
+
+For *request-response* cases there may be a need to handle timeout errors.
+
+In such case *Trigger T1* (see scenario 3 for a reference) after emitting `OracleRequest` to *Oracle O* will also emit another `OracleRequest` to a special *Timeout Oracle* which tracks timeouts. Payload will contain the timeout value.
+
+*Trigger T2* will have more complex activation filter, which also tracks store asset related to *Timeout Oracle*. When execution is started it analyzes triggering event if it got response from *Oracle O* or just awakened by *Timeout Oracle*.
+
+Obviously that this solution still relies on some Oracle but just not the original one. As we are in the blockchain nothing can happen without a documented reason or *transaction* to be more precise. So any solution which requires some error checking like that will require another Oracle, just a more trusted one.
+
+### Decisions
+
+List all decisions you and other stakeholders made
+
+### Alternatives
+
+1. Make one peer (i.e. *Leader*) directly call some external *API* when trigger submits `EmitOracleRequest` instruction. This is a more on-chain variant and allows us to create integrations with popular oracle providers on our side. However it' not so secure and performant.
+2. Implement our own IBC, make Oracle providers to run their own Iroha 2 based networks and make triggers use this IBC to retrieve data from Oracle blockchains. How exactly Oracle blockchains get data from off-chain world is out of the scope.
+
+### Concerns
+
+There are some concerns and unresolved problems:
+
+1. Payment. Typically Oracles want you to pay them. This is not a big deal for centralized solutions, because in that case payment can be done off chain. But for decentralized automatic solutions we should think how to attach payment to every `OracleRequest` event.
```

Comment:
Could payment logic be part of the trigger?

Reply:
Hm, yes, I think it can. But there are too much ways I can see:

1. Permission to withdraw the amount of money from trigger account. That way *Oracle* can withdraw the exact amount of money it wants for its computations
2. Basic pre-pay inside trigger. If there will be some change, that *Oracle* transfers it back.
3. Pay when the exact Oracle result is retrieved. Such result should contain the amount of money to pay.

I think there aren't any best solution here and every one of them is suitable for different cases. So we might want to keep this as offtop for this RFC.

Reply:
Actually thank you for the idea. I've put that in the document